### ISSUE TITLE
[BUG] Cannot increase batch size more than 1 with ZeRO-Infinity SSD offloading

### ISSUE THREAD

**Issue by jhm0104666** [microsoft/DeepSpeed#4047]
## **Describe the bug**
I cannot increase the batch size as much as stated in ZeRO-Infinity SSD offloading (arXiv '21). The configuration used in the paper is attached below. 

I ran training with facebook/opt-30b, which has similar configurations as the 50B model in the table. Due to the large similarity in the configuration, I expected that I could increase the batch size up to ~26. However, I could only use the batch size of 1; otherwise I ran into a CUDA OOM error. Moreover, I could not run larger models even with the base size of 1 (e.g., facebook/opt-175b).

FYI, I've tried a number of optimizations listed in various sources to avoid the occurrence of CUDA OOM (e.g., [link1](https://www.deepspeed.ai/docs/config-json/), [link2](https://discuss.pytorch.org/t/keep-getting-cuda-oom-error-with-pytorch-failing-to-allocate-all-free-memory/133896/6)).

![image](https://github.com/microsoft/DeepSpeed/assets/43145676/17ed876a-5023-4af6-83a1-721c90b1be99)

### Workload 
- Python script: [transformer example clm code](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py) except:
  - Commented out line 58: check_min_version(..)
  - Commented out line 583: trainer.save_model()
- Model: facebook/opt-30b
- Dataset: wikitext-2-raw-v1

### Run script
- Setting {BATCH SIZE} to a number larger than 1 gives an OOM error.
- I used various pytorch configs to reduce the CUDA OOM events, but nothing worked (e.g., PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:256).
- I set zero-related parameters to lower numbers, but nothing worked (e.g., stage3_max_live_parameters, stage3_max_reuse_distance).
```
PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:256 deepspeed --master_port 60000 --num_nodes=1 --num_gpus={NUM GPU TO USE} run_clm.py --deepspeed config.json --model_name_or_path facebook/opt-30b --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --do_train --fp16 --per_device_train_batch_size {BATCH SIZE} --learning_rate 2e-5 --num_train_epochs 1 --output_dir result --overwrite_output_dir --save_steps 0 --max_steps 4 --save_strategy "no"
```

### DeepSpeed config (config.json)
```
{
  "train_micro_batch_size_per_gpu": "auto",
  "fp16": {
    "enabled": true
  },
  "optimizer": {
    "type": "Adam"
  },
  "zero_optimization": {
    "stage": 3,
    "offload_optimizer": {
      "device": "nvme",
      "nvme_path": "NVME_DEV_TO_BE_FILLED",
      "pin_memory": true,
      "buffer_count": 4,
      "fast_init": false
    },
    "offload_param": {
      "device": "nvme",
      "nvme_path": "NVME_DEV_TO_BE_FILLED",
      "pin_memory": true,
      "buffer_count": 50,
      "buffer_size": 1e9,
      "max_in_cpu": 1e9
    },
    "overlap_comm": true,
    "contiguous_gradients": true,
    "stage3_max_live_parameters": 1e9,
    "stage3_max_reuse_distance": 1e9,
    "reduce_bucket_size": "auto",
    "stage3_prefetch_bucket_size": "auto",
    "stage3_param_persistence_threshold": "auto"
  },
  "aio": {
    "block_size": 1048576,
    "queue_depth": 32,
    "thread_count": 8,
    "single_submit": true,
    "overlap_events": true
  },
  "activation_checkpointing": {},
  "flops_profiler": {
      "enabled": true,
      "profile_step": 1,
      "module_depth": -1,
      "top_modules": 1,
      "detailed": true,
      "output_file": null
  }
}
```

## **Expected behavior**
- Could increase the batch size up to 26 => could increase only up to 1.
- Could run bigger models such as opt-175b => could not run them.

## **ds_report output**
```
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/hamin.jang/work/Profile/LLM/profile_LLM/venv/lib/python3.10/site-packages/torch']
torch version .................... 2.0.1+cu118
deepspeed install path ........... ['/home/hamin.jang/work/Profile/LLM/profile_LLM/venv/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.9.2, unknown, unknown
torch cuda version ............... 11.8
torch hip version ................ None
nvcc version ..................... 11.8
deepspeed wheel compiled w. ...... torch 2.0, cuda 11.8
```

## **Screenshots**
![image](https://github.com/microsoft/DeepSpeed/assets/43145676/59881486-b002-4a89-84d6-184d31a3b100)

## **System info (please complete the following information):**
- Intel CPU server with 1T memory
- One machine with NVIDIA A100-40g GPU (x1~4), NCCL enabled
- PCIe Gen4 x16 
- NVMe SSDs
- OS: Ubuntu 22.04 LTS
- NVIDIA driver: 515.105.01, CUDA Version: 11.8
- Python 3.10
<details>
<summary>Pip requirements</summary>

```
--extra-index-url https://download.pytorch.org/whl/cu118
accelerate==0.19.0
aiohttp==3.8.4
aiosignal==1.3.1
async-timeout==4.0.2
attrs==23.1.0
certifi==2022.12.7
charset-normalizer==2.1.1
cmake==3.25.0
datasets==2.13.1
deepspeed==0.9.2
dill==0.3.6
evaluate==0.4.0
filelock==3.9.0
frozenlist==1.3.3
fsspec==2023.5.0
hjson==3.1.0
huggingface-hub==0.14.1
idna==3.4
inquirerpy==0.3.4
Jinja2==3.1.2
lit==15.0.7
MarkupSafe==2.1.2
mpmath==1.2.1
multidict==6.0.4
multiprocess==0.70.14
networkx==3.0
ninja==1.11.1
numpy==1.24.1
packaging==23.1
pandas==2.0.2
pfzy==0.3.4
Pillow==9.3.0
prompt-toolkit==3.0.38
psutil==5.9.5
py-cpuinfo==9.0.0
pyarrow==12.0.1
pydantic==1.10.8
python-dateutil==2.8.2
pytz==2023.3
PyYAML==6.0
regex==2023.5.5
requests==2.28.1
responses==0.18.0
six==1.16.0
sympy==1.11.1
tokenizers==0.13.3
torch==2.0.1+cu118
torchaudio==2.0.2+cu118
torchvision==0.15.2+cu118
tqdm==4.65.0
transformers==4.29.2
triton==2.0.0
typing_extensions==4.4.0
tzdata==2023.3
urllib3==1.26.13
wcwidth==0.2.6
xxhash==3.2.0
yarl==1.9.2

```
</details>

## **Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?
> `deepspeed` launcher

## **Docker context**
Are you using a specific docker image that you can share?
> I didn't use docker images

**Comment by tjruwase** [comment]
@jhm0104666, the OOM occurs during the forward pass, which it is due to activation memory consumption. 
![image](https://github.com/microsoft/DeepSpeed/assets/4271600/91e6bf13-8255-40e3-9ba9-42c44d172c56)

ZeRO-Infinity only helps with parameter and optimizer memory footprint. In our evaluation we reduced activation memory footprint using [tiling](https://deepspeed.readthedocs.io/en/latest/zero3.html#memory-centric-tiling) and [activation or gradient checkpointing](https://deepspeed.readthedocs.io/en/latest/activation-checkpointing.html). The later [optimization ](https://huggingface.co/docs/transformers/v4.18.0/en/performance#gradient-checkpointing)is also available in transformers and could be worth exploring.

**Comment by jhm0104666** [comment]
@tjruwase Thanks for your valuable advise. I looked into the two optimization techniques mentioned above, but it is hard to grasp how to use the memory tiling one. Is this unavailable with my current SW setups which uses a DS json config file and HF transformer? The documentation says that DS-infinity supports memory tiling, but I couldn't find any way to configure this in DS json. Could you provide me some example usage of the technique?

**Comment by jhm0104666** [comment]
@tjruwase I've also tried activation checkpointing, but still got the same error at the same forward pass. 

Due to the following reason, I guess some mechanisms in DeepSpeed regarding data fetch/release are not working properly.
- Theoretically, the number of paramters in one layer of opt-30b is about 620M, so if FP16 is used, the size of param and activation should be 1.2GB each (2.4GB in total). Therefore, a layer in the forward pass requires {2.4GB + other temporary GPU buffers}, which should be far less than the total GPU memory capacity. 
- As far as I know, DeepSpeed offload installs pre-/post-hook at the invocation of each layer, and these hooks fetch and release data into/outof GPU memory. If these hooks work properly, I guess CUDA OOM should never happen.

Here's the config I used for activation checkpointing.
```
    "activation_checkpointing": {
        "partition_activations": true,
        "cpu_checkpointing": true,
        "contiguous_memory_optimization": true,
        "number_checkpoints": 50,
        "synchronize_checkpoint_boundary": true,
        "profile": true
    },
```

**Comment by tjruwase** [comment]
@jhm0104666, apologies for the confusion. Using DeepSpeed's [tiling](https://github.com/microsoft/Megatron-DeepSpeed/blob/ba79efc8bc39a68f38179f68792bad1f2f19199f/megatron/arguments.py#L1424-L1437) or [activation checkpointing](https://github.com/microsoft/Megatron-DeepSpeed/blob/ba79efc8bc39a68f38179f68792bad1f2f19199f/megatron/arguments.py#L1444-L1454) optimizations require changes to the model code. For the zero-infinity results, we made these changes to our [Megatron-DeepSpeed](https://github.com/microsoft/Megatron-DeepSpeed/tree/main) model code. 

I think the HF gradient checkpointing might be easier for you to start with given your software stack. Here is an [example usage](https://github.com/microsoft/DeepSpeedExamples/blob/81e88bd98be89ae1c631efb93037f4f97aa62830/applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/main.py#L255-L258) from our DS-Chat code.

**Comment by jhm0104666** [comment]
@tjruwase OK. It seems worth trying Megatron-DeepSpeed. I will try this and let you know if there's an unexpected behavior. But it seems that if I want "the full set of functionalities" such as activation checkpointing, only a limited set of models (in Megatron-DeepSpeed repo) are available. Otherwise, I should manually modifiy HF model code to support them. 

So, as an workaround, I added "--gradient_checkpoinint" option **directly into the HF arguments**, hoping that the size of activation reduces. However, I got the error below. I've ran into this type of errors many times before when configuring the deepspeed json file, and I wonder what derives this error in the first place. Can you give me some advice regarding safely configuring the deepspeed arguments? There's no documentation addressing this topic.

![image](https://github.com/microsoft/DeepSpeed/assets/43145676/2df5217f-e729-4d71-b8ec-40720f2cb121)

**Comment by jhm0104666** [comment]
The error didn't appear only when I set the following three options to zero, and temporarily succeeded to increase the batch size: _stage3_max_live_parameters, stage3_max_reuse_distance, stage3_prefetch_bucket_size_. 

However, the performance dropped as the batch size grows (especially when 8->12; see table below), and there could be so many possible causes (e.g., more recomputation, disabled stage3 options, ..). I guess DeepSpeed team have already addressed similar issues, so could you provide me some advises regarding this?

|batch size|train_steps_per_second|train_samples_per_second|
|--|--|--|
|2|0.034|0.068
|4|0.033|0.131|
|8|0.03|0.238|
|12|0.017|0.203|
|16|0.015|0.246|
|20|0.01|0.198|

**Comment by tjruwase** [comment]
> So, as an workaround, I added "--gradient_checkpoinint" option **directly into the HF arguments**, hoping that the size of activation reduces. However, I got the error below. I've ran into this error many times before when configuring the deepspeed json file, and I wonder what derives this error in the first place. Can you give me some advice regarding safely configuring the deepspeed arguments? There's no documentation addressing this topic.
> 
> ![image](https://user-images.githubusercontent.com/43145676/257040400-2df5217f-e729-4d71-b8ec-40720f2cb121.png)

This looks like a bug to me. Could you please file a separate issue with repro details? Thanks!

**Comment by tjruwase** [comment]
> The error didn't appear only when I set the following three options to zero, and temporarily succeeded to increase the batch size: _stage3_max_live_parameters, stage3_max_reuse_distance, stage3_prefetch_bucket_size_.
> 
> However, the performance dropped as the batch size grows (especially when 8->12; see table below), and there could be so many possible causes (e.g., more recomputation, disabled stage3 options, ..). I guess DeepSpeed team have already addressed similar issues, so could you provide me some advises regarding this?
> 
> batch size	train_steps_per_second	train_samples_per_second
> 2	0.034	0.068
> 4	0.033	0.131
> 8	0.03	0.238
> 12	0.017	0.203
> 16	0.015	0.246
> 20	0.01	0.198

Yes, your instincts are correct that there could be many reasons for performance reduction with increased batch size. Unfortunately, this makes performance tuning quite challenging. One approach is to first understand which execution portion is getting slower (e.g., forward, backward, etc.). You can enable `wall_clock_breakdown` in the [ds_config](https://www.deepspeed.ai/docs/config-json/#logging) to enable profiling in the logs.

**Comment by tjruwase** [comment]
Closing as OP question is answered. Please re-open if needed.
